### PR TITLE
Auto LOD fixes and improvements

### PR DESCRIPTION
**Pull request by JFonS**
Various fixes and improvements to Auto LOD generation:
* Fixed LODs for shadow meshes.
* Added a merging step before simplification. This helps with tessellated meshes that were previously left untouched. The angle difference at which edges are considered "hard" can be tweaked as an import setting.
* LODs will now start with the highest decimation possible and keep doubling (approximately) the number of triangles from there. This makes sure that very low triangle counts are included when possible.
* Given more weight to normal preservation.
* Modified MeshOptimizer to report distance-based error instead of including attributes in the reported metrics.
* Added attribute transference between the original mesh and the various LODs. Right now, only normals are taken into account,
  but it could be expanded to other attributes in the future.

I also added the `StaticRaycaster` class that can be used to cast rays against a set of static meshes. It's basically a wrapper for Embree that is used in the attribute transference code, we could eventually expose it to scripting for editor tools or other cases where fast ray intersections are needed and  meshes are all static.

### Tests
As an example, take this heavily subdivided mesh:
image

Here you can see the various LOD levels:

https://user-images.githubusercontent.com/4402304/134935099-890a1fa9-308d-400e-94d3-d5dc519dc1c2.mp4

And how they look when zooming in and out of the mesh:

https://user-images.githubusercontent.com/4402304/134719611-5730ac49-478a-4c51-bb1e-5c2349de48c8.mp4

Finally, a comparison of a single scene with auto LOD disabled and enabled:
lods

Even with the camera being relatively close, the number of primitives drops by a factor of 10, gaining almost a millisecond of GPU time. There are some small visual differences, but they are hard to notice unless you know beforehand where to look at.

### CODE REVIEW

File: editor/import/scene_importer_mesh.cpp [hunk 1]
```diff
@@ -182,67 +255,216 @@ void EditorSceneImporterMesh::generate_lods() {
...
+			lod.distance = mesh_error * scale;
 			lod.indices = new_indices;
-			print_line("Lod " + itos(surfaces.write[i].lods.size()) + " begin with " + itos(indices.size() / 3) + " triangles and shoot for " + itos(index_target / 3) + " triangles. Got " + itos(new_len / 3) + " triangles. Lod screen ratio " + rtos(lod.distance));
+			//print_line(vformat("Lod %d begin with %d triangles and shoot for %d triangles. Got %d triangles. Lod screen ratio %s.", surfaces.write[i].lods.size(), index_count / 3, index_target / 3, new_index_count / 3, vformat("%f (%f relative error)", lod.distance, mesh_error)));
 			surfaces.write[i].lods.push_back(lod);
-			index_target *= threshold;
+			index_target *= 2;
+			last_index_count = new_index_count;
+		}
+		//int new_verts_count = split_vertex_count - vertex_count;
```
Comment (fire) [inline editor/import/scene_importer_mesh.cpp:459]: Probably want to remove the print lines that aren't kept. Thoughts?

**Comment by fire** [comment]
Can you make a video of the current result of decimation?

**Comment by fire** [comment]
I tried the auto lod branch and the import seems slow. Are you aware of this?

**Comment by JFonS** [comment]
@fire I will try to get some video on monday. 

I also noticed the increased import times, but only on debug builds. Apparently Embree is much slower in debug builds than it is on optimized builds. I did some tests building just the Embree thirdparty code with the `-O3` flag while keeping the rest of the editor as a debug build, and the import times are similar to current master, or even faster in some cases.

I've been considering the idea of forcing optimized builds for Embree regardless of Godot's build target. I would only leave some way to disable it in the `SCsub` file for the module, in case we need to debug a specially nasty issue. Most bugs would likely come from Godot's code rather than Embree itself, so I think we can safely force optimized builds for it and reduce import times for any developers running Godot debug builds.

**Comment by JFonS** [comment]
I updated the PR with support for tessellated meshes and other small fixes. It's good for review and merging now.

**Comment by reduz** [comment]
image

this looks a bit odd to me, the barrel is originally smoothed, so why does it look flattened in the LOD?

**Comment by JFonS** [comment]
Because meshoptimizer is picking vertices that are in the ridge, so normals are incorrect. After the simplification step we compute new normals for faces where that happens, but those are not perfect either. However, I don't think this is really noticable in the zooming in/out video, so it should work fine.

**Comment by reduz** [comment]
Yeah but in this case, since the object is smoothed, if it cant find a proper normal it should compute a new one based on the faces connected to it but remain smooth. Here it seems a normal is generated for each face, which seems wasty to me.

**Comment by reduz** [comment]
I can understand this is difficult to do, so for this I would try to approximate by storing whether the vertex is smoothed or flat (you can detect if more than one vertex exists at the position), its not ideal but should work for the most part. If smooth, you recreate a normal from the connected faces and compare to the current one and replace if it diverges too much. If flat, you can just create a new one for each connected triangle.

**Comment by JFonS** [comment]
I can take a look at keeping those vertices flat. Honestly, it didn't cross my mind while working on it, but it shouldn't be too hard to do.

**Comment by JFonS** [comment]
I updated the PR so now it does a better job at keeping smooth vertices smooth. Here's the updated video:

https://user-images.githubusercontent.com/4402304/134935099-890a1fa9-308d-400e-94d3-d5dc519dc1c2.mp4

**Comment by fire** [comment]
This is work for a different proposal, but I found it hard to know the number and to trigger specific lod levels for verification. 

Was specifically look for cases like the barrel where the topology changes from box to a disconnected pieces.

**Comment by fire** [comment]
Can someone else confirm that blend shapes are correctly LOD'd.

**Comment by reduz** [comment]
@JFonS wow this looks _so_ much better.

**Review by reduz — APPROVED** [review]

File: editor/import/scene_importer_mesh.cpp [hunk 2]
```diff
@@ -190,67 +263,343 @@ void EditorSceneImporterMesh::generate_lods() {
...
+
+			if (new_index_count < last_index_count * 1.5f) {
+				index_target = index_target * 1.5f;
+				continue;
 			}
-			Surface::LOD lod;
-			lod.distance = mesh_error * scale;
-			if (Math::is_zero_approx(mesh_error)) {
+
+			if (new_index_count <= 0 || (new_index_count >= (index_count * 0.75f))) {
```
Comment (aaronfranke) [inline editor/import/scene_importer_mesh.cpp:409]: Is the intention of this code that the LOD generation will stop once it surpasses an upper bound of 75% of the original mesh's indices?

**Pull request merged**